feat(provider): add NextDNS

hieunlt opened the pull request:
Adds support for [NextDNS](https://nextdns.io/), resolves https://github.com/qdm12/ddns-updater/issues/321

hieunlt committed:
Add support for NextDNS

hieunlt commented:
@qdm12 can you take a look at my PR?

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,40 @@
+
+### Compulsory parameters
+
+- `"domain"` is the domain to update. For now, it must be "link-ip.nextdns.io".
```
Does it always HAVE TO be `link-ip.nextdns.io`? If so, let's hardcode it and enforce `domain` is always set to this.

hieunlt replied:
done

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,40 @@
+### Compulsory parameters
+
+- `"domain"` is the domain to update. For now, it must be "link-ip.nextdns.io".
+- `"endpoint"`
```
Should we split endpoint following the `https://link-ip.nextdns.io/<Endpoint ID>/<API GUID>` format from the issue? 🤔  It would be nice to add a bit of a description to this too.

hieunlt replied:
agree, also update the `Domain setup` section to guide users

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,40 @@
+### Optional parameters
+
+- `"ip_version"` can be `ipv4` (A records), or `ipv6` (AAAA records) or `ipv4 or ipv6` (update one of the two, depending on the public ip found). It defaults to `ipv4 or ipv6`.
+- `"ipv6_suffix"` is the IPv6 interface identifier suffix to use. It can be for example `0:0:0:0:72ad:8fbb:a54e:bedd/64`. If left empty, it defaults to no suffix and the raw public IPv6 address obtained is used in the record updating.
```
(updating to master branch)

```suggestion
- `"ipv6_suffix"` is the IPv6 interface identifier suffix to use. It can be for example `0:0:0:0:72ad:8fbb:a54e:bedd/64`. If left empty, it defaults to no suffix and the raw temporary IPv6 address of the machine is used in the record updating. You might want to set this to use your permanent IPv6 address instead of your temporary IPv6 address.
```

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,40 @@
+
+> NextDNS supports updating Linked IP via a DDNS hostname. If you're already using a DDNS service, configure your DDNS domain in the Linked IP card instead.
+
+- Create an account on the [nextdns website](https://nextdns.io/)
+- Go to your [account page](https://my.nextdns.io/), login and setup Linked IP
+- Click `Show advanced options` button and copy the endpoint from the Linked IP card
+- Update the configuration file with the endpoint
```
```suggestion
1. Create an account on the [nextdns website](https://nextdns.io/)
1. Go to your [account page](https://my.nextdns.io/), login and setup Linked IP
1. Click `Show advanced options` button and copy the endpoint from the Linked IP card
1. Update the configuration file with the endpoint
```

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,40 @@
+- Click `Show advanced options` button and copy the endpoint from the Linked IP card
+- Update the configuration file with the endpoint
+
+_See the [nextdns website](https://nextdns.io/)_
```
nit unneeded formatting

```suggestion
See the [nextdns website](https://nextdns.io/)
```

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,40 @@
+
+## Domain setup
+
+> NextDNS supports updating Linked IP via a DDNS hostname. If you're already using a DDNS service, configure your DDNS domain in the Linked IP card instead.
```
nit unneeded formatting

```suggestion
NextDNS supports updating Linked IP via a DDNS hostname. If you're already using a DDNS service, configure your DDNS domain in the Linked IP card instead.
```

qdm12 commented on internal/provider/providers/nextdns/provider.go:
```diff
@@ -0,0 +1,144 @@
+		return fmt.Errorf("%w: %s", errors.ErrOwnerWildcard, owner)
+	case !endpointRegex.MatchString(endpoint):
+		return fmt.Errorf("%w: endpoint %q does not match regex %q",
+			errors.ErrTokenNotValid, endpoint, endpointRegex)
```
`errors.ErrTokenNotValid` is not appropriate here

hieunlt replied:
created `errors.ErrEndpointNotValid`

qdm12 commented on internal/provider/providers/nextdns/provider.go:
```diff
@@ -0,0 +1,144 @@
+	}
+	defer response.Body.Close()
+
+	s, err := utils.ReadAndCleanBody(response.Body)
```
What's in the response on success? How about with the wrong endpoint?

hieunlt replied:
Tested the `link-ip.nextdns.io` endpoint:

200 OK — success, plain-text OK body (IP updated)
400 Bad Request — plain-text Bad Request; happens when endpoint_id or api_guid don't match the expected format
404 Not Found — JSON {"errors":[{"code":"notFound"}]}; happens when path segments are missing entirely

hieunlt committed:
Update docs/nextdns.md

Co-authored-by: Quentin McGaw <quentin.mcgaw@gmail.com>

hieunlt committed:
Update docs/nextdns.md

Co-authored-by: Quentin McGaw <quentin.mcgaw@gmail.com>

hieunlt committed:
Update docs/nextdns.md

Co-authored-by: Quentin McGaw <quentin.mcgaw@gmail.com>

hieunlt committed:
Apply suggestions from code review

Co-authored-by: Quentin McGaw <quentin.mcgaw@gmail.com>

hieunlt committed:
Refactor NextDNS provider settings and update documentation

- Updated the endpoint format in the documentation to specify the required structure.
- Removed the domain and owner parameters from the New function, setting defaults directly in the Provider struct.
- Simplified the validateSettings function to only check the endpoint format.

hieunlt committed:
Refactor NextDNS provider to use separate endpoint_id and api_guid parameters

- Updated the NextDNS provider to accept endpoint_id and api_guid instead of a single endpoint string.
- Enhanced validation logic to check for the presence of both parameters.
- Added unit tests for the new validation logic.
- Updated documentation to reflect the changes in parameter structure.

hieunlt committed:
Enhance error handling in NextDNS provider Update method

- Updated the Update method to handle different HTTP response statuses more explicitly.
- Added specific error messages for BadRequest responses and maintained existing error handling for other statuses.

hieunlt committed:
Add default domain for NextDNS provider when not specified

- Set the default domain to 'link-ip.nextdns.io' if the domain is empty and the provider is NextDNS.
- This change simplifies user configuration by eliminating the need for users to specify the domain explicitly.

hieunlt committed:
Update default domain for NextDNS provider to 'nextdns.io'

- Changed the default domain from 'link-ip.nextdns.io' to 'nextdns.io' in both settings and provider initialization.
- This update reflects the fixed nature of the domain for NextDNS, simplifying user configuration.

hieunlt committed:
Implement domain and owner validation for NextDNS provider

- Added mandatory parameters for `domain` and `owner` in the NextDNS provider, ensuring they are set to `nextdns.io` and `link-ip`, respectively.
- Updated the provider initialization to validate these parameters and return appropriate errors for invalid configurations.
- Enhanced unit tests to cover various scenarios for domain and owner validation.

hieunlt commented:
hi @qdm12 I've made a few adjustments based on your comments, can you take a look when you have time?

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,44 @@
+
+### Compulsory parameters
+
+- `"domain"` must be `"nextdns.io"` and `"owner"` must be `"link-ip"`.
```
1. let's stick to just using `"domain": "link-ip.nextdns.io"`. `owner` is just here for retrocompatibility
2. why have this field there if it's always a constant? Just have it as a constant in the code right? Maybe the code is too rigid to allow that, just saying this from the top of my head. If so, just change it to be `domain` only, without `owner`.

qdm12 commented on docs/nextdns.md:
```diff
@@ -0,0 +1,44 @@
+1. Copy the update URL shown — it looks like `https://link-ip.nextdns.io/abcdef/0123456789abcdef`
+1. Set `"endpoint_id"` to the first path segment (e.g. `abcdef`) and `"api_guid"` to the second path segment (e.g. `0123456789abcdef`)
+
+See the [nextdns website](https://nextdns.io/)
```
nit not necessary given instructions above 😉 

```suggestion

```

qdm12 commented:
I did push a few changes, but importantly the `domain` field is now your actual domain name you are updating, not `link-ip.nextdns.io`. The update URL is constant in the update code. The domain field should be set by the user so that ddns-updater can dns resolve it and figure out when a dns update call is needed. This is critical to avoid updating every period, even if there is no need for it. Thoughts?

Edit: also apologies for the mega late replies, I'm swamped with life and code, still there, but lagging pretty hard

hieunlt commented:
> The domain field should be set by the user so that ddns-updater can dns resolve it and figure out when a dns update call is needed. This is critical to avoid updating every period, even if there is no need for it. Thoughts?

thanks man, as a follow up I removed redundant test cases, at first I believed that `domain` is fixed from NextDNS side, that's why I hardcoded it in the first place and thank you very much for spending time explain the mechanism to me 🙏

qdm12 commented on internal/provider/providers/nextdns/provider_test.go:
```diff
@@ -11,58 +11,16 @@ import (
-		})
-	}
+	require.NoError(t, err)
+	assert.NotNil(t, provider)
```
let's assert it against an expected Provider struct, otherwise this test is a bit useless except to check it does not error
